Lib/pathlib.py

```diff
@@ -1185,10 +1186,10 @@ def expanduser(self):
         """
         if (not (self._drv or self._root) and
             self._parts and self._parts[0][:1] == '~'):
-            homedir = self._flavour.expanduser(self._parts[0])
+            homedir = self._flavour.expanduser(self)
```

Comment:
Is this change (and the one below) necessary? Seems we shouldn't ever have a `~USERNAME` anywhere other than in the first part?

We shouldn't have to format the string just to parse it again.

Reply:
Without this change, `WindowsPath('~/a:').expanduser()` results in `a:` rather than `c:/users/barney/a:`. This is because the code joined `c:/users/barney` and `a:` using `os.path.join()`!

The implemementation of `expanduser()` doesn't call `os.path.join()` - instead it concatenates the remainder as a string. This is the behaviour we want, I believe!

Reply:
Hmm, I see. Though that would suggest that `WindowsPath('~').expanduser() / "a:b"` would also create the wrong result. (Some of our join implementations probably also treat `./a:b` as a drive change, too - pretty sure I wrote one that will do it.) I'm not sure how to decide which is the bigger problem.

Actually, this only matters in the case of single character filename, which is fairly unlikely, though also incredibly unlikely for anyone to add special cased code for it.

Do/can we have a `_from_parts` call that doesn't normalise/join anything? I would have thought that would be the cheap way to do this, so maybe we do - I haven't even looked. At least in the case where we start from a parsed path, we would be best to avoid re-parsing it in case we get a different result.

Reply:
> Though that would suggest that `WindowsPath('~').expanduser() / "a:b"` would also create the wrong result.

You could disambiguate it with `WindowsPath('~').expanduser() / "./a:b"`. Pathlib removes `.` segments *only after* spliting the path with `os.path.splitroot()`.

> Do/can we have a _from_parts call that doesn't normalise/join anything? I would have thought that would be the cheap way to do this, so maybe we do - I haven't even looked. At least in the case where we start from a parsed path, we would be best to avoid re-parsing it in case we get a different result.

That's a good shout - it may be possible to use `_from_parsed_parts()` here. I'll look into it!

Reply:
Thanks for the pointer, using `_from_parsed_parts()` looks like just the ticket.